### ISSUE TITLE
Logic error that prevents correct group affinity binding of the first 8 threads

### ISSUE THREAD

**Issue by noobpwnftw** [official-stockfish/Stockfish#1312]
https://github.com/official-stockfish/Stockfish/blob/7dd1f4a7c027832889c2e20c2131bab9d64c8095/src/misc.cpp#L301

On a Windows machine with multiple processor groups, the first 8 threads are not assigned to any processor groups so they are stuck in the same group where the process were initially started.
i.e. if you have 2 groups of 64 cores, setting Threads to 128 would result in one group having 72 threads and the other one having 56 threads.

**Comment by mcostalba** [comment]
Do you have the hardware to verify a possible fix?

**Comment by joergoster** [comment]
I thought this condition would only prevent any binding with not more than 7 threads running, which seems reasonable.
If instead it would say ```if (idx < 8) return;``` then the first threads would not be assigned to any processor group.
Am I missing something?

**Comment by noobpwnftw** [comment]
@mcostalba Yes I have the hardware to verify the fix.

@joergoster "Threads" here is a ThreadPool vector which is creating threads in set() function, these threads will start to execute immediately and this condition is checked in the newly created threads, there are chances that this < 8 condition is met when the ThreadPool vector haven't grow large enough.

**Comment by joergoster** [comment]
@noobpwnftw I see. Thx for explaining.

**Comment by joergoster** [comment]
@noobpwnftw A simple 'workaround' seems to be to read the set UCI option 'Threads' instead of using Threads.size().

```
  int NumThreads = Options["Threads"];

  if (NumThreads < 8)
      return;
```
If you want to give it a try, don't forget to include uci.h in misc.cpp.

**Comment by CoffeeOne** [comment]
I would like to comment on this issue: I have a 40 core Intel machine running Windows, with HT turned on it is seen as 2 processor groups with 40 threads each. So I can make tests. What should be tested?
What is the problem? I also do not really understand that if (Threads.size() < 8) condition. Is it only because of fishtest? I have never seen a machine with more than 64 threads on fishtest, so for what is the condition then?

**Comment by vondele** [comment]
I've pushed a branch here https://github.com/vondele/Stockfish/tree/rebind maybe @CoffeeOne could give this a try ? Given his running test here http://tests.stockfishchess.org/tests/view/5a3693480ebc590ccbb8be5a it would be sufficient to test if gets the same speedup from this branch as from his fix. For the case he is testing, the solutions should be the same.

**Comment by CoffeeOne** [comment]
@vondele I tested my "simple fix", with many 1 thread tasks: I started games with 1 Thread with concurrency 39 with cutechess-cli on the 4-way Xeon E7-4870 (fishtest simulation): I saw immediately, that this is not good, all threads were only running on one physical CPU, meaning one CPU had 100% CPU load and 3 CPUs 0% CPU load.

So I fully support your solution, please make a pull-request for "rebind"
Then I ran another last benchmark between
Engine 1: Branch rebind
Engine 2: Latest Master
Engine 3: Latest Master with simple fix (outcommented if)

Branch "rebind" is in the same speed range than simple-fix, so +1!
Build Tester: 1.4.7.0
Windows Server 2012 R2 (Version 6.3, Build 0, 64-bit Edition)
Intel(R) Xeon(R) CPU E7- 4870  @ 2.40GHz
HyperThreading Enabled: Yes
Command Line: bench 1024 40 26 default depth
Tests per Build: 5
                Engine# (NPS)                     Speedup     Sp     Conf. 95%    S.S.
1  (49 351 529,4  ) ---> 2  (43 699 189,8  ) ---> 12,935%  2 641 862,6    Yes       No 
1  (49 351 529,4  ) ---> 3  (48 997 237,6  ) --->  0,723%  960 292,8    No        No

**Comment by vondele** [comment]
@CoffeeOne thanks for testing, will prepare a pull request.

**Comment by mcostalba** [comment]
Fixed by 1c50d8cbf554733c0db6ab. Closing.
